kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java

```diff
@@ -20,104 +35,319 @@
+          < cachedMessageDescriptorRetensionSeconds) {
+        return cachedDescriptor.getDescriptor();
+      }
+    }
+
+    String bufOwner = bufDefaultOwner;
+    String bufRepo = "";
+
+    String bufOwnerRepoInfo = bufOwnerRepoByProtobufMessageName.get(fullyQualifiedTypeName);
+
+    if (bufOwnerRepoInfo != null) {
+      String[] parts = bufOwnerRepoInfo.split("/");
+      if (parts.length != 2) {
+        log.error("Cannot parse Buf owner and repo info from {}, make sure it is in the 'owner/repo' format",
+            bufOwnerRepoInfo);
+      } else {
+        bufOwner = parts[0];
+        bufRepo = parts[1];
+      }
+    } else {
+      String[] parts = fullyQualifiedTypeName.split("\\.");
+
+      if (parts.length == 0) {
+        log.warn("Cannot infer Buf repo name from type {}", fullyQualifiedTypeName);
+      } else {
+        bufRepo = parts[0];
+      }
+    }
+
+    log.info("Get descriptor from Buf {}/{}@{}", bufOwner, bufRepo, fullyQualifiedTypeName);
+
+    Descriptor descriptor = bufClient.getDescriptor(bufOwner, bufRepo, fullyQualifiedTypeName);
+
+    cachedDescriptor = new CachedDescriptor(currentDate, descriptor);
+    cachedMssageDescriptorMap.put(fullyQualifiedTypeName, cachedDescriptor);
+
+    return descriptor;
+  }
+
+  private String parse(byte[] value, Descriptor descriptor) throws IOException {
+    DynamicMessage protoMsg = DynamicMessage.parseFrom(
+        descriptor,
+        new ByteArrayInputStream(value));
+    byte[] jsonFromProto = ProtobufSchemaUtils.toJson(protoMsg);
+    return new String(jsonFromProto, utf8Charset);
+  }
+
   @Override
   public ProducerRecord<byte[], byte[]> serialize(String topic,
-                                                  @Nullable String key,
-                                                  @Nullable String data,
-                                                  @Nullable Integer partition) {
-    return this.schemaRegistryAwareRecordSerDe.serialize(topic, key, data, partition);
+      @Nullable String key,
+      @Nullable String data,
+      @Nullable Integer partition) {
+    if (data == null) {
+      return new ProducerRecord<>(topic, partition, Objects.requireNonNull(key).getBytes(), null);
+    }
+
+    ProtoSchema protoSchema = protoSchemaFromTopic(topic);
+    if (protoSchema == null) {
+      throw new RuntimeException("Could not infer proto schema from topic " + topic);
+    }
+
+    Descriptor descriptor = getDescriptor(protoSchema.getFullyQualifiedTypeName());
+    if (descriptor == null) {
+      throw new RuntimeException("Could not get descriptor for " + protoSchema.getFullyQualifiedTypeName());
+    }
+
+    DynamicMessage.Builder builder = DynamicMessage.newBuilder(descriptor);
+
+    DynamicMessage message;
+    try {
+      JsonFormat.parser().merge(data, builder);
+      message = builder.build();
+    } catch (InvalidProtocolBufferException e) {
+      throw new RuntimeException("Failed to merge record for topic " + topic, e);
```

Comment:
I bet `InvalidProtocolBufferException` is an IOException sub class, so you can just use `new UncheckedIOException(e)`